utils/file.go

```diff
@@ -112,3 +113,25 @@ func CopyDir(src string, dst string) (err error) {
 
 	return
 }
+
+var SizeLimitExceeded = errors.New("Size limit exceeded")
+
+type LimitedReaderWithError struct {
+	limitedReader *io.LimitedReader
+	maxBytes      int64
+}
+
+func NewLimitedReaderWithError(reader io.Reader, n int64) *LimitedReaderWithError {
+	return &LimitedReaderWithError{
+		limitedReader: &io.LimitedReader{R: reader, N: n},
+		maxBytes:      n,
+	}
+}
+
+func (l *LimitedReaderWithError) Read(p []byte) (int, error) {
+	n, err := l.limitedReader.Read(p)
+	if l.limitedReader.N <= 0 && err == io.EOF {
```

Comment:
In this specific case I think this is fine and will work as intended, but considering `NewLimitedReaderWithError` is laid out as a general-purpose utility that could end up being used elsewhere... this is from the `io.Reader` docs:

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.

So -- in theory, for some `io.Reader` implementation -- we could see an EOF here even before there was any attempt to go over the limit. In that case this `Read` method should return the original EOF error and not the `SizeLimitExceeded` one. Right?

Some of Go's internal code solves this by setting the limit one byte *higher* than what's actually the intended maximum, then checking just the `N` value: https://cs.opensource.google/go/x/mod/+/refs/tags/v0.5.1:zip/zip.go;l=658-670

Reply:
Here `N` is the number of [**remaining** bytes](https://pkg.go.dev/io#LimitedReader), it will only be `<=0` after we've exceeded the limit.

Reply:
Do you mean the check should be `<` instead of `<=`?

Reply:
> Do you mean the check should be < instead of <=?

Take a look at the link. `<=` is right but `err == io.EOF` is wrong, *and* the initial `N` value should be set to `n + 1` instead of `n`.

> Here N is the number of [remaining bytes](https://pkg.go.dev/io#LimitedReader), it will only be `<=0` after we've exceeded the limit.

This is very subtly incorrect. It will only be `<= 0` after exceeding the limit *or exactly at the limit*. If the initial `N` value is set to `n + 1`, *then* it will only be `<= 0` after exceeding the limit. And in that case checking the error value is unnecessary.

Reply:
I think I understand what you mean:

If `l.limitedReader.N` starts off as the MaxByteLimit (it reduces as data is read from the reader), 

then the conditional

`if l.limitedReader.N <= 0 && err == io.EOF `

is incorrect, because `l.limitedReader.N == 0 && err == io.EOF` means that the file/stream was exactly MaxByteLimit size and so we've hit the EoF, and this should not be a `SizeLimitExceeded` error.



>And in that case checking the error value is unnecessary.

This could potentially hide an underlying error. https://cs.opensource.google/go/go/+/refs/tags/go1.8:src/io/io.go;l=429-439

On line 438, the std library returns the error from the underlying `Read`, after setting the N value. 
The EOF is only returned on the next call to the limited reader's `Read`.

If we don't check that `err = io.EOF` we could be masking an underlying error. Of course, we are still in a `SizeLimitExceeded` state, but I wanted to return any underlying error instead of overwriting it, and the next call to the `Read` would still raise the `SizeLimitExceeded` error as expected.



Interestingly, all of this is moot when https://github.com/golang/go/commit/fd6ef06296b55a234d4fd5ebf5dd3e5bfd8f4e25 is released in some future version of go :)

Reply:
Yup, exactly.

Reply:
I don't see why we can't just use our own `LimitedReader` implementation instead of trying to override the existing one just to return a custom error. It's near about the same lines of code anyways. We just need to return our custom error instead of `EOF`.

```go
func (l *LimitedReader) Read(p []byte) (n int, err error) {
	if l.N <= 0 {
		err := l.Err
		if err == nil {
			err = // OUR CUSTOM ERROR HERE
		}
		return 0, err
	}
	if int64(len(p)) > l.N {
		p = p[0:l.N]
	}
	n, err = l.R.Read(p)
	l.N -= int64(n)
	return
}
```

Reply:
>I don't see why we can't just use our own LimitedReader implementation instead of trying to override the existing one just to return a custom error. It's near about the same lines of code anyways. We just need to return our custom error instead of EOF.

As I said before, the code you quoted was merged into Go master just yesterday and will released in 1.19, at which time we can use it directly. 
For the version of `io.LimitedReader` in our supported version of Go (1.18), which I looked at, it seemed wrapping made more sense than duplicating, because I think the wrapper I've written is shorter, easier to understand, and expresses the intent better (all subjective, sure).

Reply:
0/5 on it really, but I don't really think writing our own LimitedReader implementation is that complicated. Your conversation with Juho made me think that there maybe subtle edge-cases in trying to implement a correct wrapper. So I thought that just copying the implementation until Go 1.19 is available should do for now. But it's a non-blocker, so approving it.

Reply:
>Your conversation with Juho made me think that there maybe subtle edge-cases in trying to implement a correct wrapper. 

This is true, but I think the wrapper is good now. I think Juho might prefer copying the Go standard library version, so I'll leave it to him to take the final call.

Reply:
I'm 0/5 on what's the right thing to do and in this specific use-case it doesn't seem to matter at all. Fine with either option.